spec/abstractops.html

```diff
@@ -857,6 +857,50 @@ <h1>ISO 8601 grammar</h1>
           MinuteSecond
           `60`
 
+      TimeHourNotValidMonth : one of
+          `00` `13` `14` `15` `16` `17` `18` `19` `20` `21` `23`
+
+      TimeHourNotThirtyOneDayMonth : one of
+          `02` `04` `06` `09` `11`
```

Comment:
Can we find a way to make this work that _doesn't_ require such bizarre one-off productions, e.g. early errors?

Reply:
I'm not sure if early errors would be appropriate for this, since there's no separate parsing and evaluation step as there is with source text. I'd be happy to investigate it though, did you have particular language in mind?

Another idea, maybe something like this would work?
```
TimeHourMinuteBasicFormatNotAmbiguous :
    TimeHour TimeMinute but not DateSpecMonthDay
```

One advantage of doing it explicitly with one-off productions like this, though, is that it corresponds more closely to the valid strings generator in `test/validStrings.mjs` which is really helpful for debugging the grammar.

If we don't figure out a better way to write this before the plenary, I do think we can present this next week anyway and rewrite it in a subsequent editorial change if need be.

Reply:
[ParseText](https://tc39.es/ecma262/#sec-parsetext) (emphasis mine):
> Attempt to parse _sourceText_ using _goalSymbol_ as the [goal symbol](https://tc39.es/ecma262/#sec-context-free-grammars), and **analyse the parse result for any [early error](https://tc39.es/ecma262/#early-error) conditions**. Parsing and [early error](https://tc39.es/ecma262/#early-error) detection may be interleaved in an [implementation-defined](https://tc39.es/ecma262/#implementation-defined) manner.

So it's valid to define and use early errors in productions that correspond with specialized grammars, and in fact this is already used for [regular expressions](https://tc39.es/ecma262/#sec-patterns-static-semantics-early-errors) and [JSON.parse](https://tc39.es/ecma262/#sec-object-initializer-static-semantics-early-errors).

Reply:
Sounds good, what's your opinion on whether this is necessary in order to present this PR at plenary next week? It seems to me that we could do it later as an editorial change.

Reply:
→ #1984